design/keycloak.x/operator.md

```diff
@@ -0,0 +1,128 @@
+
+
+## Features
+
+---
+**NOTE**
+
+The primary target of the operator is to make it easy to achieve production grade installations of Keycloak.X.
+
+---
+
+### Configuring Keycloak deployment
+
+The operator will use a CRD representing the Keycloak installation. The CRD will expose the following configuration options:
+* custom Keycloak image; default base image will be used if not specified
+* Git configuration to fetch static business objects configuration (see below)
+* manual horizontal and vertical scaling
+* pod and node affinity, toleration
+* SSL config, truststore
+
+Since most of the configuration options will come from the Keycloak.X distribution itself, the CRD will also expose appropriate fields for passing any distribution related options to the container, like database connection details (obviously without any credentials), SPIs configuration, etc.
+
+### Configuring Keycloak business objects using Kubernetes resources
+
+The new operator will expose two main ways of configuring business objects in Keycloak (e.g.: Realms, Roles, Clients, etc.) in addition to the built-in Dynamic configuration through the REST API/UI console:
+* Static file based configuration stored in Git to enable GitOps operations
+* Static configuration through Kubernetes API
+
+Static configuration will be strictly read-only, therefore two-way syncing is not going to be needed.
+
+Static configuration is going to provide an immutable and cloud native way for managing business objects in Keycloak that can be easily moved between environments (e.g. dev, stage, prod) in a predictable manner. This feature will leverage the new Store.X which enables federation of the configuration from multiple sources (static/dynamic) by re-architecting the storage layer.
+
+#### Static configuration through Git
+
+The `Keycloak` CRD will enable defining a specific commit (identified by an hash) in a Git repository containing the static configuration in the form of JSON/YAML files. To update the configuration the user will simply change a commit hash in a `Keycloak` CR and the operator will roll out the new configuration to all pods.
+
+#### Static configuration through Kubernetes API
+
+The operator will leverage dedicated CRD(s), initially, there will be only one `Realm` CRD directly translated from Keycloak's [RealmRepresentation](https://github.com/keycloak/keycloak/blob/c7134fd5390d7c650b3dfd4bd2a2855157042271/core/src/main/java/org/keycloak/representations/idm/RealmRepresentation.java). A Realm includes all subresources. As a result, it is going to be possible to configure every object in Keycloak through this CR even though for some of them it won't be recommended (e.g. Users). To implement this, the operator will simply translate the CRs to YAML files and mount them to Keycloak pods, again leveraging Store.X.
```

Comment:
It is quite weird that we have the Realm CRD but we don't recommend configuring some key aspects like users. In practice, you are favoring managing users (and some others) through the admin console, right?

IMO, we should *provision* realms and then rely on the administration console. However, I see some space for improvement to enable some key capabilities/configuration when provisioning realms such as:

* Pre-configure for FAPI
* Pre-configured for a specific registration/authentication flow
* Brokering using social providers
* Any additional realm capability/configuration that otherwise the user would need to configure by himself.

We would have some realm templates that users can choose from to provision the realm.

Manage realms (fully) using CRs does not look user-friendly for me.

Reply:
> We would have some realm templates that users can choose from to provision the realm.

this is a great idea!

Reply:
+1, i really like the idea of templates.

Reply:
The templates functionality will basically have to come from the Store.X. Operator won't be actively involved in this. To my understanding, it will indeed work as you're suggesting. A "template" will be the static config (stored either in git or as a CR) while some config will be dynamic (modifiable through REST API / Admin Console) thanks to different store types aggregation. Correct @hmlnarik @mhajas ?

Reply:
Realm templates makes little sense to me. This should rather be addressed by improving the RealmRepresentation and making it easier to configure things than to have to rely on templates/examples.